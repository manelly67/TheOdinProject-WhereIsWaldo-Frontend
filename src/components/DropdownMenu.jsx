import { useState } from "react";
import { getNormalizedCoords } from "./coord";

// CREAR un componente escribir etiqueta con base a una respueta correct/incorrect
const DropdownMenu = (props) => {
  const [classes, setClasses] = useState(props.dropdownMenu);
  const [selected, setSelected] = useState(
    props.imgCharacters ? props.imgCharacters[0].name : null
  );

  function displayMenu(event) {
    event.stopPropagation();
    toggleOpenClose();
  }

  function toggleOpenClose() {
    if (classes === props.dropdownMenu) {
      openMenu();
    } else {
      closeMenu();
    }
  }

  function openMenu() {
    setClasses(props.visible);
  }

  function closeMenu() {
    setClasses(props.dropdownMenu);
  }

  function getObject(selected) {
    let filtered = props.imgCharacters.filter((e) => e.name === selected);
    if (filtered) {
      props.setSelectedChar(filtered[0]);
    }
  }

  function submitChar(event) {
    console.log("funcion submit char");

    event.stopPropagation();
    getNormalizedCoords(
      props.tagginCoords,
      props.coords,
      props.W,
      props.H,
      props.setNormalizeCoords
    );
    getObject(selected);
    // DESDE AQUI SE LLAMARA LA FUNCION PARA EL BACKEND
    // con el objeto personaje, las coordenadas normalizadas, el jugador, y el id del juego o partida
    closeMenu();
    props.setClickImg(false);
  }

  return (
    <>
      {props.clickImg ? (
        props.imgCharacters.length > 0 ? (
          <>
            <div className={`${props.placeMenu}`}>
              <button
                className={`${props.visible}`}
                onClick={(event) => {
                  displayMenu(event);
                }}
              >
                TAGS
              </button>
              <div
                role="menu"
                className={`${classes}`}
                aria-hidden={classes === props.visible ? false : true}
              >
                <select
                  name="character"
                  onClick={(event) => {
                    event.stopPropagation();
                  }}
                  onChange={(event) => setSelected(event.target.value)}
                >
                  {props.imgCharacters.map((e) => {
                    return (
                      <option
                        key={e.id}
                        value={e.name}
                        onClick={(event) => {
                          event.stopPropagation();
                        }}
                      >
                        {e.name}
                      </option>
                    );
                  })}
                </select>
                <button
                  onClick={(event) => {
                    submitChar(event);
                  }}
                >
                  submit
                </button>
              </div>
            </div>
          </>
        ) : null
      ) : null}
    </>
  );
};

export default DropdownMenu;
