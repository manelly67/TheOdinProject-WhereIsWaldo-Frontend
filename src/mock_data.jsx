const mock_data_1 = {
  game_id: "generate_id",
  startedAt: new Date(),
  picture: {
    src_image:
      "https://res.cloudinary.com/dwlqplcgt/image/upload/v1743608952/find_the_charaters_in_the_galactic_city_pcuwu3.png",
    characters: [
      { id: "char-1", name: "Wally", found: false, x: 0, y: 0 },
      { id: "char-2", name: "R2D2", found: false, x: 0, y: 0 },
      {
        id: "char-3",
        name: "Courage the Cowardly Dog",
        found: false,
        x: 0,
        y: 0,
      },
    ],
  },
};

const mock_data_2 = {
  game_id: "generate_id",
  startedAt: new Date(),
  picture: {
    src_image:
      "https://res.cloudinary.com/dwlqplcgt/image/upload/v1743608911/find_the_lucky_clover_mqkdmu.png",
    characters: [
      { id: "char-4", name: "Four Leaf Clover", found: false, x: 0, y: 0 },
    ],
  },
};

const mock_player = {
  player: {
    id: "ec335d1b-98c8-42b5-aa45-94099a7b041d",
    playername: "ANONIMOUS",
    sessionId: "k8WaVzsac5EBe7mfw0ORAVoFf53ptUh-",
  },
};

const mock_game = {  // game database not the oject response
  id: "f59e2335-8d8c-431d-9432-f91ea8513b92",
  startedAt: "2025-04-11 22:16:02.423",
  finishedAt: null,
  timeRecord: {},
  playerId: "5290bdec-150a-4c48-b740-d791ea5e9ad8",
  pictureId: "img-1",
  status: "GAMING",
  targets: [
    { x: 0, y: 0, id: "char-1", name: "Wally", found: false },
    { x: 0, y: 0, id: "char-2", name: "R2D2", found: false },
    {
      x: 0,
      y: 0,
      id: "char-3",
      name: "Courage the Cowardly Dog",
      found: false,
    },
  ],
  timeInSeconds: null,
};

export { mock_data_1, mock_data_2, mock_player, mock_game };
