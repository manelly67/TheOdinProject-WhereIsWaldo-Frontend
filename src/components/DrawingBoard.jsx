import { Link } from "react-router-dom";
import { useState, useEffect, useCallback } from "react";
import { useLocation } from "react-router-dom";
import ToggleTheme from "./ToggleTheme";
import TargetingBox from "./TargetinBox";
import DropdownMenu from "./DropdownMenu";
import TagTheChar from "./TagTheChar";
import Message from "./Message";
import AskForName from "./AskForName";
import styles from "../styles/Box.module.css";
import { moveToCoord, clickInsideImg } from "./coord";
import { urlAddresses } from "../assets/urlAdresses";

let didInitImg1 = false;
let didInitImg2 = false;

const DrawingBoard = () => {
  const { box, placeMenu, dropdownMenu, visible } = styles;
  const [gameName, setGameName] = useState(null);
  const titleDiv = document.querySelector("title");
  if (titleDiv) {
    titleDiv.textContent = gameName;
  }
  const location = useLocation();

  const [game, setGame] = useState(null);
  const [messageObj, setMessageObj] = useState(null);
  const [isInTop, setIsInTop] = useState(false);
  const [dontAskAgain, setDontAskAgain] = useState(false);
  const imgSource = game === null ? null : game.picture.src_image;
  const imgId = game === null ? null : game.picture.id_image;
  const imgCharacters = game === null ? null : game.picture.characters;
  const score = game === null ? null : formatScore(game.timeRecord);
  
  const width = window.innerWidth;
  const height = window.innerHeight;
  const [coords, setCoords] = useState({ x: 0, y: 0 });
  const [endcoords, setEndcoords] = useState({ x: 0, y: 0 });
  const [W, setW] = useState(0);
  const [H, setH] = useState(0);
  const [tagginCoords, setTagginCoords] = useState({ x: 0, y: 0 });
  const [clickImg, setClickImg] = useState(false);
  const [stop, setStop] = useState(false);

  const getData = useCallback(async (url) => {
    try {
      await fetch(url, {
        method: "POST",
        credentials: "same-origin",
        headers: {
          "Content-Type": "application/json",
          Connection: "keep-alive",
        },
      })
        .then((res) => res.json())
        .then((data) => {
          if (data.game) {
            const { game } = data;
            setGame(game);
          }
        })
        .catch((err) => {
          console.log(err);
        });
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  const initBoard = useCallback(async () => {
  
    if (location.state !== null) {
      const { player, gameName } = location.state;
      setGameName(gameName);

      if (gameName !== null) {
        if (gameName === "Waldo In The Galactic City") {
          const url = `${urlAddresses.createGame}/img-1/${player.id}`;
          if (!didInitImg1) {
            didInitImg1 = true;
            await getData(url);
          }
        }
        if (gameName === "Oh! Waldo is not here") {
          const url = `${urlAddresses.createGame}/img-2/${player.id}`;
          if (!didInitImg2) {
            didInitImg2 = true;
            await getData(url);
          }
        }
      }
    }
  }, [location.state, getData]);

  const getImgCoord = useCallback(() => {
    const imageRef = document.getElementById(`${gameName}`);
    if (imageRef) {
      const rect = imageRef.getBoundingClientRect();
      setCoords({
        x: Number(rect.left.toFixed(10)),
        y: Number(rect.top.toFixed(10)),
      });
      setEndcoords({
        x: Number(rect.right.toFixed(10)),
        y: Number(rect.bottom.toFixed(10)),
      });
      setW(Number(rect["width"].toFixed(4)));
      setH(Number(rect["height"].toFixed(4)));
    }
  }, [gameName]);

  const updateData = useCallback(async (url) => {
    try {
      const response = await fetch(url, { method: "GET" });
      const responseData = await response.json();
      if (responseData.game) {
        const { game } = responseData;
        setGame(game);
      }
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  const checkIsInTopTen = useCallback(async (game) => {
    const url = `${urlAddresses.topTEN}/${game.picture.id_image}`;
    try {
      const response = await fetch(url);
      const responseData = await response.json();
      if (responseData.top10) {
        const { top10 } = responseData;
        top10.forEach((r) => {
          if (r) {
            if (r.id === game.game_id) {
              setIsInTop(true);
            }
          }
        });
      }
    } catch (error) {
      alert("Something was wrong. try again later");
      console.log(error);
    }
  }, []);

  useEffect(() => {
    if (game === null) {
      initBoard();
    }
  }, [game, initBoard]);

  useEffect(() => {
    if (location.state !== null) {
      const { player, gameName } = location.state;
      if (gameName !== null && player !== null) {
        if (gameName === "Waldo In The Galactic City") {
          const url = `${urlAddresses.createGame}/img-1/${player.id}`;
          updateData(url);
        }
        if (gameName === "Oh! Waldo is not here") {
          const url = `${urlAddresses.createGame}/img-2/${player.id}`;
          updateData(url);
        }
      }
    }
  }, [location.state, updateData]);

  useEffect(() => {
    if (imgSource !== null) {
      getImgCoord();
    }
  }, [gameName, imgSource, getImgCoord]);

  useEffect(() => {
    const temp = clickInsideImg(tagginCoords, coords, endcoords);
     setClickImg(temp);
  }, [tagginCoords, coords, endcoords]);

  useEffect(() => {
    const f = document.getElementById("box");
    if (f) {
      if (tagginCoords.x === 0 && tagginCoords.y === 0) {
        f.style.transform = `translateY(${tagginCoords.y}px)`;
        f.style.transform += `translateX(${tagginCoords.x}px)`;
      } else {
        f.style.transform = `translateY(${tagginCoords.y - 5}px)`;
        f.style.transform += `translateX(${tagginCoords.x - 15}px)`;
      }
    }
  }, [tagginCoords]);

  useEffect(() => {
    switch (game === null) {
      case true:
        break;
      case false:
        switch (game.status === "ENDED") {
          case true:
            setStop(true);
            if (game.player.name_player === "ANONIMOUS") {
              if (!dontAskAgain) {
                checkIsInTopTen(game);
              }
            }
            break;
          case false:
            switch (game.status === "GAMING") {
              case true: {
                document.addEventListener(
                  "click",
                  (ev) => {
                    let { movetoX, movetoY } = moveToCoord(
                      ev.clientX,
                      ev.clientY,
                      ev.pageX,
                      ev.pageY
                    );
                    setTagginCoords({ x: movetoX, y: movetoY });
                  },
                  false
                );
                return () => {
                  document.removeEventListener(
                    "click",
                    (ev) => {
                      let { movetoX, movetoY } = moveToCoord(
                        ev.clientX,
                        ev.clientY,
                        ev.pageX,
                        ev.pageY
                      );
                      setTagginCoords({ x: movetoX, y: movetoY });
                    },
                    false
                  );
                };
              }
              default:
                break;
            }
        }
    }
  }, [game, dontAskAgain, checkIsInTopTen]);

  function formatScore(score) {
    if (score) {
      const { days, hours, minutes, seconds } = score;
      let d = days ? `${days} days` : "";
      let h = hours ? `${hours} hours` : "";
      let m = minutes ? `${minutes} minutes` : "";
      let s = seconds ? `${seconds} seconds` : "";
      return `${d} ${h} ${m} ${s}`;
    }
  }

  return (
    <>
      <div className="bar">
        <>
          {!stop ? (
            <>
              <div id="box" className={`${box}`}>
                <div>
                  <TargetingBox />
                </div>
                <div>
                  <DropdownMenu
                    clickImg={clickImg}
                    setClickImg={setClickImg}
                    game={game}
                    setGame={setGame}
                    coords={coords}
                    W={W}
                    H={H}
                    tagginCoords={tagginCoords}
                    setTagginCoords={setTagginCoords}
                    placeMenu={placeMenu}
                    dropdownMenu={dropdownMenu}
                    visible={visible}
                    imgCharacters={imgCharacters}
                    setMessageObj={setMessageObj}
                  />
                </div>
              </div>
            </>
          ) : null}

          <div>
            <ToggleTheme theme="dark" />
          </div>

          <div className="setOfLinks">
            <Link to="/" style={{ fontSize: "1.3rem" }}>
              HOME
            </Link>
            <Link
              to={"/top_ten"}
              state={{ img_id: imgId, gameName: gameName }}
              style={{ fontSize: "1.3rem" }}
            >
              TOP TEN
            </Link>
            {!game ? null : !game.timeInSeconds ? null : (
              <>
                <div>
                  <p> Your SCORE:</p>
                  <p style={{ width: "80px" }}>{score}</p>
                </div>
              </>
            )}
          </div>
        </>
      </div>
      <div>
        {!imgSource ? (
          <>
            <div>
              <p>loading</p>
              <p>
                Dear user, this is a study project, please wait 1 minute for the
                server to wake up.
              </p>
            </div>
          </>
        ) : (
          <>
            <img
              id={gameName}
              src={imgSource}
              alt={gameName}
              className=""
              width={width * 0.75}
              height={
                width * 0.8 * 0.65 < height ? width * 0.8 * 0.65 : height * 0.95
              }
            ></img>
          </>
        )}
      </div>
      {imgCharacters ? (
        <TagTheChar imgCharacters={imgCharacters} coords={coords} W={W} H={H} />
      ) : null}

      {messageObj ? (
        <Message messageObj={messageObj} setMessageObj={setMessageObj} />
      ) : null}

      {isInTop ? (
        <AskForName
          game={game}
          setIsInTop={setIsInTop}
          setDontAskAgain={setDontAskAgain}
        />
      ) : null}
    </>
  );
};

export default DrawingBoard;
