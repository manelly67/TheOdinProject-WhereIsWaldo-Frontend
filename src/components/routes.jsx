import App from "../App";
import ErrorPage from "./Error_page";
import DrawingBoard from "./DrawingBoard";
import TopTen from "./TopTen";

const routes = [
  {
    index: true,
    path: "/",
    element: <App />,

    errorElement: <ErrorPage />,
  },
  {
    path: "board",
    element: <DrawingBoard />,
  },
  {
    path: "top_ten",
    element: <TopTen />,
  },
];

export default routes;
