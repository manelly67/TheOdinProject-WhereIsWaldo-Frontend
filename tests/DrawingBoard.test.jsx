import { describe, it, expect, vi } from "vitest";
import { render, screen} from "@testing-library/react";
import {
  MemoryRouter,
  createBrowserRouter,
  RouterProvider,
} from "react-router-dom";
import userEvent from "@testing-library/user-event";


import routes from "../src/components/routes.jsx";
import {mock_data_1, mock_data_2} from "../src/mock_data.jsx";
import DrawingBoard from "../src/components/DrawingBoard";
import DropdownMenu from "../src/components/DropdownMenu.jsx";

describe("simple use of mock fuction", () => {
  it("mock function", () => {
    const fn = vi.fn();
    fn("hello world");

    expect(fn.mock.calls[0]).toEqual(["hello world"]);

    const market = {
      getApples: () => 100,
    };

    const getApplesSpy = vi.spyOn(market, "getApples");
    market.getApples();

    expect(getApplesSpy.mock.calls.length).toEqual(1);
    expect(market.getApples()).toBe(100);
  });
});

describe("display DrawingBoard", () => {
  it("display the page when location.state is null", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/board"]}>
        <DrawingBoard />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
    expect(
      screen.getByText(
        /Dear user, this is a study project, please wait 1 minute for the server to wake up./i
      )
    ).toBeInTheDocument();
  });

  it("display drawing board when location.state is not null", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();

    const buttons = screen.getAllByRole("button");
    const button = buttons[1]; // choose the button for galactic city

    await user.click(button);

    expect(
      screen.getByAltText(/Waldo In The Galactic City/i)
    ).toBeInTheDocument();
  });

  it("render behavior when the click is inside the image", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();

    const buttons = screen.getAllByRole("button");
    const button = buttons[1];
    await user.click(button);  // display the image
    
    // when click Img is true
    render(
      <DropdownMenu
        clickImg={true}
        imgCharacters={mock_data_1.picture.characters}
      />
    );
    expect(screen.getByText(/tags/i)).toBeInTheDocument();
  });

  it("render behavior when the click is outside the image", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();
    const buttons = screen.getAllByRole("button");
    const button = buttons[1];
    await user.click(button);  // display the image

    // when click Img is false
    render(
      <DropdownMenu
        clickImg={false}
        imgCharacters={mock_data_1.picture.characters}
      />
    );
    expect(screen.queryByText(/tags/i)).not.toBeInTheDocument();


  });
});
