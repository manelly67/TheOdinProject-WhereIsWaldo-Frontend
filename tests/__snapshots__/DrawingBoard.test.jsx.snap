// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`display DrawingBoard > display the page when location.state is null 1`] = `
<div>
  <div
    class="bar"
  >
    <div
      class="_box_3f9b6c"
      id="box"
      style="transform: translateY(0px)translateX(0px);"
    >
      <div>
        <svg
          role="presentation"
          style="width: 1.5rem; height: 1.5rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H15V21H19C20.1 21 21 20.1 21 19V15H19M19 3H15V5H19V9H21V5C21 3.9 20.1 3 19 3M5 5H9V3H5C3.9 3 3 3.9 3 5V9H5M5 15H3V19C3 20.1 3.9 21 5 21H9V19H5V15M7 11H9V13H7V11M11 11H13V13H11V11M15 11H17V13H15V11M7 7H9V9H7V7M11 7H13V9H11V7M15 7H17V9H15V7M7 15H9V17H7V15M11 15H13V17H11V15M15 15H17V17H15V15Z"
            style="fill: currentColor;"
          />
        </svg>
      </div>
      <div />
    </div>
    <div>
      <button
        class="themeButton"
      >
        <img
          alt="theme-light-dark"
          class="iconImg"
          height="30px"
          src="/src/assets/theme-light-dark.png"
          width="30px"
        />
      </button>
    </div>
    <div
      class="setOfLinks"
    >
      <a
        data-discover="true"
        href="/"
        style="font-size: 1.3rem;"
      >
        HOME
      </a>
    </div>
  </div>
  <div>
    <div>
      <p>
        loading
      </p>
      <p>
        Dear user, this is a study project, please wait 1 minute for the server to wake up.
      </p>
    </div>
  </div>
</div>
`;
